### ISSUE TITLE
Exception: NoMethodError: undefined method `gsub' for false:FalseClass

### ISSUE THREAD

**Issue by deeco** [chef/knife-windows#227]
Hello,

When trying to bootstrap a windows server using Windows Powershell I get the following error.

I am using the command : knife bootstrap windows winrm localhost -VV

I can see the client list if I run knife client list
Also I have read something about the validator.pem but this appears to be in the right directory.

![image](https://cloud.githubusercontent.com/assets/10785989/7632184/6c83a48e-fa42-11e4-8506-bbe21f3507b6.png)

```
PS C:\chef\config> knife bootstrap windows winrm localhost -VV
INFO: Using configuration from C:/chef/config/knife.rb

Waiting for remote response before bootstrap.DEBUG: Adding localhost
Enter your password:
DEBUG: :session => :init
DEBUG: :relay_to_servers => echo . & echo Response received.
DEBUG: :relayed => localhost
DEBUG: localhost => :run_command
DEBUG: localhost[6CD4454D-36AD-481C-8ABB-8B36A0E7D425] => :shell_open
DEBUG: localhost[6CD4454D-36AD-481C-8ABB-8B36A0E7D425] => :run_command[echo . & echo Response received.]
DEBUG: localhost[6CD4454D-36AD-481C-8ABB-8B36A0E7D425] => :command_cleanup[echo . & echo Response received.]
DEBUG: localhost[6CD4454D-36AD-481C-8ABB-8B36A0E7D425] => :shell_close
DEBUG: :command_complete => localhost with return code 0
DEBUG: command complete on localhost
DEBUG: command complete on All Servers
DEBUG: :relay_output_from_backend => ["localhost", ". \r\n"]
localhost .
DEBUG: :relay_output_from_backend => ["localhost", "Response received.\r\n"]
localhost Response received.
DEBUG: Exit code found: 0
DEBUG: :unbind => :connection
DEBUG: localhost => :unbind
DEBUG: :unbind_backend => localhost
DEBUG: :session => :close
Remote node responded after 0.2 minutes.
Bootstrapping Chef on localhost
DEBUG: Looking for bootstrap template in C:/opscode/chef/embedded/lib/ruby/gems/2.0.0/gems/knife-windows-0.8.4/lib/chef
knife/bootstrap
DEBUG: Found bootstrap template in C:/opscode/chef/embedded/lib/ruby/gems/2.0.0/gems/knife-windows-0.8.4/lib/chef/knife
bootstrap
ERROR: knife encountered an unexpected error
This may be a bug in the 'bootstrap windows winrm' knife command or plugin
Please collect the output of this command with the `-VV` option before filing a bug report.
Exception: NoMethodError: undefined method `gsub' for false:FalseClass
PS C:\chef\config>
```

**Comment by ac-rmenezes** [comment]
> On May 14, 2015, at 10:07 AM, Deeco <notifications@github.com> wrote:
> 
> Hello,
> 
> When trying to bootstrap a windows server using Windows Powershell I get the following error.
> 
> I am using the command : knife bootstrap windows winrm localhost -VV
> 
> I can see the client list if I run knife client list
> Also I have read something about the validator.pem but this appears to be in the right directory.
> 
>  <https://cloud.githubusercontent.com/assets/10785989/7632184/6c83a48e-fa42-11e4-8506-bbe21f3507b6.png>
> PS C:\chef\config> knife bootstrap windows winrm localhost -VV
> INFO: Using configuration from C:/chef/config/knife.rb
> 
> Waiting for remote response before bootstrap.DEBUG: Adding localhost
> Enter your password:
> DEBUG: :session => :init
> DEBUG: :relay_to_servers => echo . & echo Response received.
> DEBUG: :relayed => localhost
> DEBUG: localhost => :run_command
> DEBUG: localhost[6CD4454D-36AD-481C-8ABB-8B36A0E7D425] => :shell_open
> DEBUG: localhost[6CD4454D-36AD-481C-8ABB-8B36A0E7D425] => :run_command[echo . & echo Response received.]
> DEBUG: localhost[6CD4454D-36AD-481C-8ABB-8B36A0E7D425] => :command_cleanup[echo . & echo Response received.]
> DEBUG: localhost[6CD4454D-36AD-481C-8ABB-8B36A0E7D425] => :shell_close
> DEBUG: :command_complete => localhost with return code 0
> DEBUG: command complete on localhost
> DEBUG: command complete on All Servers
> DEBUG: :relay_output_from_backend => ["localhost", ". \r\n"]
> localhost .
> DEBUG: :relay_output_from_backend => ["localhost", "Response received.\r\n"]
> localhost Response received.
> DEBUG: Exit code found: 0
> DEBUG: :unbind => :connection
> DEBUG: localhost => :unbind
> DEBUG: :unbind_backend => localhost
> DEBUG: :session => :close
> Remote node responded after 0.2 minutes.
> Bootstrapping Chef on localhost
> DEBUG: Looking for bootstrap template in C:/opscode/chef/embedded/lib/ruby/gems/2.0.0/gems/knife-windows-0.8.4/lib/chef
> knife/bootstrap
> DEBUG: Found bootstrap template in C:/opscode/chef/embedded/lib/ruby/gems/2.0.0/gems/knife-windows-0.8.4/lib/chef/knife
> bootstrap
> ERROR: knife encountered an unexpected error
> This may be a bug in the 'bootstrap windows winrm' knife command or plugin
> Please collect the output of this command with the -VV option before filing a bug report.
> Exception: NoMethodError: undefined method `gsub' for false:FalseClass
> PS C:\chef\config>
> 
> —
> Reply to this email directly or view it on GitHub <https://github.com/chef/knife-windows/issues/227>.
> 

Hi Deeco,


Had this one before, but not trying to bootstrap the same workstation I’m working on (where you obviously need to have chef-client already set). Not sure if it *can* be done or if it is a good practice. By not sure I mean I don’t know - some wisdom from the experts would benefit both of us.

Some questions:
Using a starter kit? Can the node/workstation (!) reach the internet? Windows version?

IF you want to give it a go anyway, try to follow the entire process (maybe again), and see if You can get this done:

install ruby and rubygems (Im on ruby 2.0.0p481, gem 2.0.14)
run gem install knife-windows

On PS:
 
winrm quickconfig -q
winrm set winrm/config/winrs '@{MaxMemoryPerShellMB="300"}' 
winrm set winrm/config '@{MaxTimeoutms=“3000000"}’ 
winrm set winrm/config/service '@{AllowUnencrypted="true"}' 
winrm set winrm/config/service/auth '@{Basic="true"}'


netsh advfirewall firewall add rule name="WinRM 5985" protocol=TCP dir=in localport=5985 action=allow
netsh advfirewall firewall add rule name="WinRM 5986" protocol=TCP dir=in localport=5986 action=allow
net stop winrm
sc config winrm start=auto
net start winrm

Bootstrap:

knife bootstrap windows winrm 127.0.0.1 --winrm-user Administrator --winrm-password ‘YourPassWord' --node-name WindowsNodeName



Regards
Rafael

**Comment by ac-rmenezes** [comment]
> On May 14, 2015, at 10:40 AM, Rafael Menezes <rmenezes@avenuecode.com> wrote:
> 
> 
>> On May 14, 2015, at 10:07 AM, Deeco <notifications@github.com <mailto:notifications@github.com>> wrote:
>> 
>> Hello,
>> 
>> When trying to bootstrap a windows server using Windows Powershell I get the following error.
>> 
>> I am using the command : knife bootstrap windows winrm localhost -VV
>> 
>> I can see the client list if I run knife client list
>> Also I have read something about the validator.pem but this appears to be in the right directory.
>> 
>>  <https://cloud.githubusercontent.com/assets/10785989/7632184/6c83a48e-fa42-11e4-8506-bbe21f3507b6.png>
>> PS C:\chef\config> knife bootstrap windows winrm localhost -VV
>> INFO: Using configuration from C:/chef/config/knife.rb
>> 
>> Waiting for remote response before bootstrap.DEBUG: Adding localhost
>> Enter your password:
>> DEBUG: :session => :init
>> DEBUG: :relay_to_servers => echo . & echo Response received.
>> DEBUG: :relayed => localhost
>> DEBUG: localhost => :run_command
>> DEBUG: localhost[6CD4454D-36AD-481C-8ABB-8B36A0E7D425] => :shell_open
>> DEBUG: localhost[6CD4454D-36AD-481C-8ABB-8B36A0E7D425] => :run_command[echo . & echo Response received.]
>> DEBUG: localhost[6CD4454D-36AD-481C-8ABB-8B36A0E7D425] => :command_cleanup[echo . & echo Response received.]
>> DEBUG: localhost[6CD4454D-36AD-481C-8ABB-8B36A0E7D425] => :shell_close
>> DEBUG: :command_complete => localhost with return code 0
>> DEBUG: command complete on localhost
>> DEBUG: command complete on All Servers
>> DEBUG: :relay_output_from_backend => ["localhost", ". \r\n"]
>> localhost .
>> DEBUG: :relay_output_from_backend => ["localhost", "Response received.\r\n"]
>> localhost Response received.
>> DEBUG: Exit code found: 0
>> DEBUG: :unbind => :connection
>> DEBUG: localhost => :unbind
>> DEBUG: :unbind_backend => localhost
>> DEBUG: :session => :close
>> Remote node responded after 0.2 minutes.
>> Bootstrapping Chef on localhost
>> DEBUG: Looking for bootstrap template in C:/opscode/chef/embedded/lib/ruby/gems/2.0.0/gems/knife-windows-0.8.4/lib/chef
>> knife/bootstrap
>> DEBUG: Found bootstrap template in C:/opscode/chef/embedded/lib/ruby/gems/2.0.0/gems/knife-windows-0.8.4/lib/chef/knife
>> bootstrap
>> ERROR: knife encountered an unexpected error
>> This may be a bug in the 'bootstrap windows winrm' knife command or plugin
>> Please collect the output of this command with the -VV option before filing a bug report.
>> Exception: NoMethodError: undefined method `gsub' for false:FalseClass
>> PS C:\chef\config>
>> 
>> —
>> Reply to this email directly or view it on GitHub <https://github.com/chef/knife-windows/issues/227>.
>> 
> 
> Hi Deeco,
> 
> 
> Had this one before, but not trying to bootstrap the same workstation I’m working on (where you obviously need to have chef-client already set). Not sure if it *can* be done or if it is a good practice. By not sure I mean I don’t know - some wisdom from the experts would benefit both of us.
> 
> Some questions:
> Using a starter kit? Can the node/workstation (!) reach the internet? Windows version?
> 
> IF you want to give it a go anyway, try to follow the entire process (maybe again), and see if You can get this done:
> 
> install ruby and rubygems (Im on ruby 2.0.0p481, gem 2.0.14)
> run gem install knife-windows
> 
> On PS:
>  
> winrm quickconfig -q
> winrm set winrm/config/winrs '@{MaxMemoryPerShellMB="300"}' 
> winrm set winrm/config '@{MaxTimeoutms=“3000000"}’ 
> winrm set winrm/config/service '@{AllowUnencrypted="true"}' 
> winrm set winrm/config/service/auth '@{Basic="true"}'
> 
> 
> netsh advfirewall firewall add rule name="WinRM 5985" protocol=TCP dir=in localport=5985 action=allow
> netsh advfirewall firewall add rule name="WinRM 5986" protocol=TCP dir=in localport=5986 action=allow
> net stop winrm
> sc config winrm start=auto
> net start winrm
> 
> Bootstrap:
> 
> knife bootstrap windows winrm 127.0.0.1 --winrm-user Administrator --winrm-password ‘YourPassWord' --node-name WindowsNodeName
> 
> 
> 
> Regards
> Rafael
> 
> 


Later catch… run your commands *from* chef-repo directory.


Regards
Rafael

**Comment by deeco** [comment]
Ran commands from both config and chef-repo directory. output below appears to just timeout.
![knife1](https://cloud.githubusercontent.com/assets/10785989/7691386/ec9206bc-fdb1-11e4-90b9-03c0791c024c.PNG)
![knife2](https://cloud.githubusercontent.com/assets/10785989/7691387/ec9491ca-fdb1-11e4-9c05-0b41bd8162b9.PNG)
![knife3](https://cloud.githubusercontent.com/assets/10785989/7691415/1102a04c-fdb2-11e4-8c38-3ebe1cf62185.PNG)

**Comment by deeco** [comment]
Any ideas ?

**Comment by ac-rmenezes** [comment]
> On May 14, 2015, at 10:40 AM, Rafael Menezes <rmenezes@avenuecode.com> wrote:
> 
> already set). Not sure if it *can* be done or if it is a good practice. By not sure I mean I don’t know - some wisdom from the experts 

I’m much closer to be sure now, but wouldn’t move a finger to be completely. 
> Any ideas ?
> 

Just one: go bootstrap somebody else’s machine. : )

There’s no point in loopback bootstraping unless you’re routing some vagrant fwd voodoo.



Rafael

**Comment by Vasu1105** [comment]
@decco - NoMethodError: undefined method `gsub' for false:FalseClass  this error only comes when there is something wrong with your validtor_key setting. Could you please check C:/chef/config/knife.rb ? and check that the validator.pem is available on that location. 

Could you please share knife.rb setting with us ?

**Comment by deeco** [comment]
knife.rb, should the node = computername ?

I have edited company to companyname,  also if i run knife client list , i get valid nodes....

log_level                :info
log_location             STDOUT
node_name                'username'
client_key               'c:/chef/config/username.pem'
validation_client_name   'companyname-validator'
validation_key           'c:/chef/companyname-validator'
chef_server_url          'https://vcac-chef/organizations/companyname'
syntax_check_cache_path  'c:/chef/config/syntax_check_cache'

**Comment by Vasu1105** [comment]
Is your keys are present on C:/chef location ?
I assume its not typo validation_key should be  'c:/chef/companyname-validator.pem'

**Comment by deeco** [comment]
Yes Keys present in location C:/Chef. 

Also validation key was typo , = to following: validation_key 'c:/chef/companyname-validator.pem'

**Comment by alexpop** [comment]
I was getting the same error.
After reading https://github.com/chef/knife-windows/issues/227#issuecomment-103906661, I checked my knife.rb and found the two validator lines commented out. This is because I started doing validatorless bootstrapping of Linux nodes recently.
https://www.chef.io/blog/2015/04/16/validatorless-bootstraps/

After uncommenting 'validation_client_name' and 'validation_key' in knife.rb, I was able to bootstrap again. Would be great to see support for validatorless windows bootstrapping.

Thanks @Vasu1105

**Comment by btm** [comment]
@alexpop I'll be releasing a knife-windows 1.0.0.beta.0 today. I'll get #224 in it for validatorless bootstrapping support so we can start testing that.

We might want to make a patch to catch and raise for 0.8.x for this validation_key issue too.

**Comment by deeco** [comment]
Currently still getting error when running 
"knife bootstrap windows winrm 127.0.0.1 --winrm-user Administrator --winrm-password ‘YourPassWord' --node-name WindowsNodeName"

Any ideas

![knife](https://cloud.githubusercontent.com/assets/10785989/7833806/79f63234-0461-11e5-806a-f5daea55a5d2.PNG)

**Comment by ac-rmenezes** [comment]
@Deeco,


Can you run 

telnet 127.0.0.1 5985

on this machine and post back the output?


Rafael

**Comment by deeco** [comment]
Should have mentioned that i had tried this already but ran again , and all successful

**Comment by ac-rmenezes** [comment]
> On May 28, 2015, at 10:36 AM, Deeco <notifications@github.com> wrote:
> 
> Should have mentioned that i had tried this already but ran again , and all successful
> 
> —
> Reply to this email directly or view it on GitHub <https://github.com/chef/knife-windows/issues/227#issuecomment-106312870>.
> 
@Deeco

We need to have some standards and a objective to work this out. 

What exactly are you trying to do?

Are you using a starter kit? No? Then what/how/where?

Following some tutorial? If yes, can you point which one?

Can the node/workstation (!) reach the internet? 

Versions (Windows, Knife, Chef, etc)?


Regards
Rafael

**Comment by btm** [comment]
Should be fixed via #235, #239. Stay tuned for a beta release of knife-windows 1.0.0. We could back-port this to 0.8, but there's so much in 1.0 that's worth it.
